### DISCUSSION TITLE
Discussion: Allow and can detect private extension methods inside non-static classes

### DISCUSSION THREAD

**Discussion by kekyo** [dotnet/csharplang#2571] [Language Ideas]
# Discussion: Allow and can detect private extension methods inside non-static classes

## Summary
[summary]: #summary

Support private extension methods inside non-static classes. Specifically, allow locally usable extension methods.

## Motivation
[motivation]: #motivation

Currently, the extension methods allowed inside only static classes even so if we need it for locally.

## Detailed design
[design]: #detailed-design

For example:

```csharp
// Standard (non-static) class.
public class Foo
{
  // It uses only inside for Foo class scope, but can't use in current C# version.
  private static IEnumerable<T> Bar<T>(this IEnumerable<T> enumerable) =>
    // Some calculation...
  
  public void Baz()
  {
    // Need to use extension
    foreach (var value in list.
      Distinct().
      Bar().
      Where(v => v.Name.StartsWith("...").
      OrderBy(v => v.Name))
    {
      // ...
    }
  }
}
```

Will change the extension rules:
1. Apply static method with `this` keyword at first argument.
2. If we use it from outside class scope, the class need to apply static.

**Comment by yaakov-h** [comment]
Another approach could be by relaxing CS1109 to allow extension methods inside nested static classes.

**Comment by svick** [comment]
This sounds very similar to https://github.com/dotnet/roslyn/issues/16271, which was moved to https://github.com/dotnet/csharplang/issues/301.

**Comment by Denis535** [comment]
It would allow all C# developers to write more readable fluent code. Please do it soon.

**Comment by kekyo** [comment]
And I feel it has no side effect for language spec.

**Comment by poke** [comment]
I would like this very much. Very often, you would want to use a more fluent syntax for a rather specific thing but justifying a public extension method being placed in a separate class feels too much. By having the ability to declare local extension methods, this could be scoped very nicely.

This is very similar to #301 but “extension methods everywhere” feels like a very huge thing (justifying its “likely never” milestone) while this particular thing feels very doable since it has very little effect on other things.

**Comment by kekyo** [comment]
My intention is same as you. We can expand syntactic rules for C#, but I feel it's very hard work. So I made it simplicity and understandable for conflicting.
